File: DotNet/Turmerik/DriveExplorer/DriveItem.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Turmerik.Helpers;

namespace Turmerik.DriveExplorer
{
    public enum OfficeFileType
    {
        Word = 1,
        Excel,
        PowerPoint
    }

    public enum FileType
    {
        PlainText = 1,
        Document,
        Image,
        Audio,
        Video,
        Code,
        Binary,
        ZippedFolder
    }

    public class DriveItem<TDriveItem>
        where TDriveItem : DriveItem<TDriveItem>
    {
        public DriveItem()
        {
        }

        public DriveItem(
            DriveItem<TDriveItem> src,
            int depth = 0)
        {
            Idnf = src.Idnf;
            Name = src.Name;
            DisplayName = src.DisplayName;
            IsFolder = src.IsFolder;
            FnWoExtn = src.FnWoExtn;
            FnExtn = src.FnExtn;
            IsRootFolder = src.IsRootFolder;
            PrIdnf = src.PrIdnf;
            FileType = src.FileType;
            OfficeFileType = src.OfficeFileType;
            IsTextFile = src.IsTextFile;
            IsImageFile = src.IsImageFile;
            IsVideoFile = src.IsVideoFile;
            IsAudioFile = src.IsAudioFile;

            if (depth > 0)
            {
                int childrenDepth = depth - 1;

                SubFolders = src.SubFolders?.Select(
                    item => item.CreateFromSrc<TDriveItem>(null, childrenDepth)).ToList();

                FolderFiles = src.FolderFiles?.Select(
                    item => item.CreateFromSrc<TDriveItem>(null, 0)).ToList();
            }
            else if (depth < 0)
            {
                SubFolders = src.SubFolders;
                FolderFiles = src.FolderFiles;
            }
        }

        public string Idnf { get; set; }
        public string Name { get; set; }
        public string DisplayName { get; set; }
        public bool? IsFolder { get; set; }
        public string FnWoExtn { get; set; }
        public string FnExtn { get; set; }
        public bool? IsRootFolder { get; set; }

        public string PrIdnf { get; set; }

        public FileType? FileType { get; set; }
        public OfficeFileType? OfficeFileType { get; set; }
        public bool? IsTextFile { get; set; }
        public bool? IsImageFile { get; set; }
        public bool? IsVideoFile { get; set; }
        public bool? IsAudioFile { get; set; }

        public List<TDriveItem>? SubFolders { get; set; }
        public List<TDriveItem>? FolderFiles { get; set; }
    }

    public class DriveItem : DriveItem<DriveItem>
    {
    }

    public class DriveItem<TDriveItem, TData> : DriveItem<TDriveItem>
        where TDriveItem : DriveItem<TDriveItem, TData>
    {
        public DriveItem()
        {
        }

        public DriveItem(DriveItem<TDriveItem, TData> src, int depth = 0) : base(src, depth)
        {
            Data = src.Data;
        }

        public DriveItem(DriveItem<TDriveItem> src, int depth = 0) : base(src, depth)
        {
        }

        public TData Data { get; set; }
    }
}

File: ParcelWs/libs/trmrk-mkscripts-parcelws-sync-behavior-testapp/src/trmrk/drive-item.ts
export enum FileType {
  plainText = 1,
  document,
  image,
  audio,
  video,
  code,
  binary,
  zippedFolder,
}

export enum OfficeFileType {
  word = 1,
  excel,
  powerPoint,
}

export interface DriveItemCore {
  idnf: string | null | undefined;
  prIdnf: string | null | undefined;
  csId: string | null | undefined;
  prCsId: string | null | undefined;
  prPath: string | null | undefined;

  name: string;
  displayName: string | null | undefined;

  isFolder?: boolean | null | undefined;
  isRootFolder?: boolean | null | undefined;

  fileType?: FileType | null | undefined;
  officeFileType?: OfficeFileType | null | undefined;
  isTextFile?: boolean | null | undefined;
  isImageFile?: boolean | null | undefined;
  isVideoFile?: boolean | null | undefined;
  isAudioFile?: boolean | null | undefined;

  creationTime?: Date | null | undefined;
  lastWriteTime?: Date | null | undefined;
  lastAccessTime?: Date | null | undefined;
}

export interface DriveItem extends DriveItemCore {
  subFolders: DriveItem[];
  folderFiles: DriveItem[];
}
